### PR TITLE
Implement benchmarks for broadcasting QO types

### PR DESCRIPTION
**Pull request by apkille**
Integrates very simple benchmarking into CI, specifically for bringing QO.jl up to speed with the SciML interface ([#404](https://github.com/qojulia/QuantumOptics.jl/issues/298)). Note, at this current moment, this is **not** me reserving the following bounty #407. I am merely submitting this PR to benchmark work in merged PRs (#404, https://github.com/qojulia/QuantumOpticsBase.jl/pull/172). There is still a substantial amount of work to be done in migrating other QO benchmarks to this suite.

I should also mention: to add support for solving  SDEs with QO types, I defined the method `wiener_randn!` from DiffEqNoiseProcess.jl, which meant adding it as a dependency (this is a requirement for custom array types https://github.com/SciML/StochasticDiffEq.jl/pull/579#event-13828699414).

### CODE REVIEW

**Comment by apkille** [comment]
Ah, StochasticDiffEqs hasn't released a new version yet with my changes

**Comment by apkille** [comment]
Converting to draft until SDE.jl bumps version, which should be soon as discussed on slack.

**Comment by apkille** [comment]
@Krastanov pinging for review. The performance tracking test is failing because the benchmark result of this branch cannot be compared with master. It should work for future PRs if we merge this. Correct me if I'm wrong.

File: src/stochastic_base.jl [hunk 1]
```diff
@@ -104,3 +105,5 @@ function _check_const(op)
     end
     nothing
 end
+
+DiffEqNoiseProcess.wiener_randn!(rng::AbstractRNG,rand_vec::T) where {T<:Union{Bra,Ket,Operator}} = randn!(rng,rand_vec.data)
```
Comment (Krastanov) [inline src/stochastic_base.jl:109]: Is this a bugfix for something?

File: benchmark/benchmarks.jl [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    master!(drho, rho, p, t) = timeevolution.dmaster_h!(drho, Hobj, [Jobj], [Jdag], rates, rho, copy(obj))
+    stoch_master!(drho, rho, p, t) = timeevolution.dmaster_h!(drho, Hsobj, [Jobj], [Jdag], rates, rho, copy(obj))
+    prob = SDEProblem(master!, stoch_master!, obj, (t₀, t₁))
+end
+
+for dim in (1//2, 20//1, 50//1, 100//1)
+    for prob in zip(("schroedinger", "master"), (:(bench_schroedinger), :(bench_master)))
+        name, bench = (prob[1], prob[2])
+        # benchmark solving ODE problems on data of QO types
+        SUITE[name]["pure"][string(dim)] = @benchmarkable solve(eval($bench)($dim; pure=true), DP5(); save_everystep=false)
```
Comment (Krastanov) [inline benchmark/benchmarks.jl:102]: I am a bit worried that `eval` is part of the benchmark. Is there any change if instead you do `solve(prob) setup=(prob=eval(...))`?

File: Project.toml [hunk 3]
```diff
@@ -34,7 +35,7 @@ OrdinaryDiffEq = "5, 6"
 QuantumOpticsBase = "0.3, 0.4, 0.5"
 RecursiveArrayTools = "2, 3"
 Reexport = "0.2, 1.0"
-StochasticDiffEq = "6"
+StochasticDiffEq = "6, 6.68.0"
```
Comment (Krastanov) [inline Project.toml:38]: this should probably just be one single entry `6.68` if you want to refuse `6.67`

File: Project.toml [hunk 4]
```diff
@@ -6,6 +6,7 @@ version = "1.1.1"
 Arpack = "7d9fca2a-8960-54d3-9f78-7d1dccf2cb97"
 DiffEqBase = "2b5f629d-d688-5b77-993f-72d75c75574e"
 DiffEqCallbacks = "459566f4-90b8-5000-8ac3-15dfb0a30def"
+DiffEqNoiseProcess = "77a26b50-5914-5dd7-bc55-306e6241c503"
```
Comment (Krastanov) [inline Project.toml:9]: this requires a compat bound (I am surprised Aqua did not catch this... oh, we do not have Aqua in QuantumOptics.jl...)

File: benchmark/benchmarks.jl [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+using LinearAlgebra
+using PkgBenchmark
+
+
+const SUITE = BenchmarkGroup()
+
+prob_list = ("schroedinger", "master", "stochastic_schroedinger", "stochastic_master")
+for prob in prob_list
+    SUITE[prob] = BenchmarkGroup([prob])
+    for type in ("pure", "custom")
```
Comment (Krastanov) [inline benchmark/benchmarks.jl:14]: May we use "QO types" and "Base array types" instead of pure and custom? These confused me for a second

**Review by Krastanov — CHANGES_REQUESTED** [review]
Sorry for the slow review. This is very valuable work, thank you! Do you mind making the following small changes?

File: src/stochastic_base.jl (hunk 1 above)
Comment (apkille) [reply src/stochastic_base.jl:109]: Yes, I had to define this method to enable SDEs to be solved on QO types. I didn't see an easy PR fix to submit to SciML so I defined it here. If this is too sketchy to you, let me know and I can do some more digging to work around this problem.

File: Project.toml (hunk 4 above)
Comment (apkille) [reply Project.toml:9]: 😅 I'll submit an issue...

**Comment by apkille** [comment]
@krastanov pinging for review.

**Comment by Krastanov** [comment]
Thank you! Great to finally have benchmarks as part of CI.

**Pull request merged**